### DISCUSSION TITLE
fpdi stripping links generated by dompdf

### DISCUSSION THREAD

**Discussion by mkindred** [dompdf/dompdf#2960] [Q&A] [answered]
I'm using dompdf to generate PDFs that were previously generated by fpdf. I still need to use fpdi to merge all the individual PDFs into a single document. I was 99% done before I found that all the links that dompdf renders are stripped out by fpdi.

It seems that fpdi likes URIs that are embedded within the /Annot object, rather than into their own linked /Action object.

So fpdi allows the following link (generated in fpdf):
```
7 0 obj
<</Type /Annot /Subtype /Link /Rect [156.30 222.34 285.24 214.34] /Border [0 0 0] /A <</S /URI /URI (https://example.com/)>>>>
endobj
```

But  it doesn't allow (generated in dompdf):
```
14 0 obj
<< /Type /Annot
/Subtype /Link
/A 15 0 R
/Border [0 0 0]
/H /I
/Rect [ 144.8595 231.3043 277.8247 240.0737 ]
>>
endobj
15 0 obj
<< /Type /Action
/S /URI
/URI (https://example.com/)
>>
endobj
```

Although I've worked with fpdf and fpdi for a long time, I don't know much of about PDF standards. Is one of these link methods deprecated? I fear that moving away from fpdi isn't easy to do right now, so I need to determine whether I need to patch dompdf or fpdi to get these links working.

ETA: After a bit more probing, I suspect fpdi is removing _all_ links when merging files, regardless of the Annot vs Action tags mentioned above. I was confused because most of my previous pages were created in fpdf directly, and so they bypassed fpdi.

**Comment by mkindred** [comment]
FYI I'm using dompdf 2.0.0 in Drupal 9 via the entity_print module, fpdi 2.3.6, and fpdf 1.84.

**Comment by bsweeney** [comment, marked as answer]
It looks like this is by design? At least based on this (very old) post: https://www.sitepoint.com/community/t/tcpdf-and-fdpi-issue-with-links-in-pdf/71497. Since that's a very old post some things about import may have changed.

An enterprising user did create an extension to maintain links: https://gist.github.com/andreyvit/2020422

**Reply by mkindred** [reply]
That explains why I had such a hard time attempting to find the parsing of objects while debugging fpdi (content stream). I think I'll look into replacing fpdi with ghostscript.
